constraint-solver/src/solver/exhaustive_search.rs

```diff
@@ -85,23 +85,28 @@ fn find_unique_assignment_for_set<T: FieldElement, V: Clone + Hash + Ord + Eq +
     rc: impl RangeConstraintProvider<T, V> + Clone,
     bus_interaction_handler: &impl BusInteractionHandler<T>,
 ) -> Result<Option<BTreeMap<V, T>>, Error> {
-    match get_all_possible_assignments(variables.iter().cloned(), &rc)
-        .filter(|assignments| {
-            !constraint_system.is_assignment_conflicting(assignments, &rc, bus_interaction_handler)
-        })
-        .exactly_one()
-    {
-        Ok(assignments) => Ok(Some(assignments)),
-        Err(mut iter) => {
-            if iter.next().is_some() {
-                // There are at least two assignments, so there is no unique assignment.
-                Ok(None)
-            } else {
-                // No assignment satisfied the constraint system.
-                Err(Error::ExhaustiveSearchError)
-            }
+    let mut assignments =
+        get_all_possible_assignments(variables.iter().cloned(), &rc).filter_map(|assignments| {
+            constraint_system
+                .derive_more_assignments(assignments, &rc, bus_interaction_handler)
+                .ok()
+        });
+    let Some(first_assignments) = assignments.next() else {
+        // No assignment satisfied the constraint system.
+        return Err(Error::ExhaustiveSearchError);
+    };
+    // Intersect all assignments.
+    // A special case of this is that only one of the possible assignments satisfies the constraint system,
+    // but even if there are multiple, they might agree on a subset of their assignments.
+    let mut result = first_assignments;
```

Comment:
@Schaeff I tried to turn this into a `scan`, but I don't think it is what I want! I just want a fold with an early exit, but `scan` returns an iterator instead of the final state. LMK if you have a suggestion :)

Reply:
True! Apparently, there is even try_reduce, but it is not stable.